### ISSUE TITLE
Mio build broken on Haiku, no kpoll/kqueue

### ISSUE THREAD

**Issue by kallisti5** [tokio-rs/mio#1472]
```
   Compiling parking_lot v0.9.0
   Compiling hyper v0.12.35
   Compiling aho-corasick v0.7.15
   Compiling addr2line v0.14.0
   Compiling idna v0.1.5
   Compiling idna v0.2.0
   Compiling mime_guess v2.0.3
   Compiling futures-cpupool v0.1.8
   Compiling bytes v0.4.12
   Compiling mio v0.6.23
error[E0432]: unresolved imports `self::unix::Events`, `self::unix::Selector`
 --> /boot/home/.cargo/registry/src/github.com-1ecc6299db9ec823/mio-0.6.23/src/sys/mod.rs:5:5
  |
5 |     Events,
  |     ^^^^^^ no `Events` in `sys::unix`
6 |     Io,
7 |     Selector,
  |     ^^^^^^^^ no `Selector` in `sys::unix`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0432`.
error: could not compile `mio`
```

**Comment by Thomasdezeeuw** [comment]
We don't really support Haiku. Is there are support for `epoll` or `kqueue`?

**Comment by kallisti5** [comment]
Nope. No support for either. We follow a pretty strict posix model

**Comment by kallisti5** [comment]
*actually* it looks like mio 0.8.0 just works.  The error above is on 0.6.23

```
/Data/Code/mio> cargo build
   Compiling libc v0.2.88
   Compiling log v0.4.11
   Compiling cfg-if v0.1.10
   Compiling mio v0.8.0 (/Data/Code/mio)
    Finished dev [unoptimized + debuginfo] target(s) in 2.63s
/Data/Code/mio> cargo test
   Compiling env_logger v0.6.2
   Compiling rand v0.4.6
   Compiling mio v0.8.0 (/Data/Code/mio)
    Finished test [unoptimized + debuginfo] target(s) in 5.95s
     Running target/debug/deps/mio-753c3d872fa80196

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running target/debug/deps/aio-ddb6b9e09505d866

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running target/debug/deps/close_on_drop-6b57ddb8ed3d603e

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running target/debug/deps/events-87c06669cab39f2d

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running target/debug/deps/interest-5f7b72ee86b2086e

running 4 tests
test add ... ok
test bit_or ... ok
test fmt_debug ... ok
test is_tests ... ok

test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running target/debug/deps/poll-27c9e6654b4bc8ce

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running target/debug/deps/registering-9d474f45d2be8765

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running target/debug/deps/regressions-c6f11f684f3ee713

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running target/debug/deps/tcp-0f512657bc44fcb5

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running target/debug/deps/tcp_listener-0f8dda3615537eed

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running target/debug/deps/tcp_socket-7044b995f8501b4c

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running target/debug/deps/tcp_stream-97680294a5416030

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running target/debug/deps/udp_socket-2ac819aa49a263c1

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running target/debug/deps/unix_datagram-b6c07085797ca71d

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running target/debug/deps/unix_listener-757d9ceb365003bb

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running target/debug/deps/unix_pipe-c589c7bae56441bb

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running target/debug/deps/unix_stream-08968abdd23a4436

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running target/debug/deps/waker-6870a9f6e4095fa0

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running target/debug/deps/win_named_pipe-ba42cb257be0764f

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

   Doc-tests mio

running 21 tests
test src/event/events.rs - event::events::Events (line 20) ... ignored
test src/event/events.rs - event::events::Events::capacity (line 100) ... ok
test src/event/events.rs - event::events::Events::clear (line 164) ... ignored
test src/event/events.rs - event::events::Events::is_empty (line 114) ... ok
test src/event/events.rs - event::events::Events::iter (line 128) ... ignored
test src/event/events.rs - event::events::Events::with_capacity (line 86) ... ok
test src/event/events.rs - event::events::Iter (line 55) ... ignored
test src/event/source.rs - event::source::Source (line 41) ... ignored
test src/interest.rs - interest::Interest::add (line 61) ... ok
test src/interest.rs - interest::Interest::remove (line 77) ... ok
test src/lib.rs - guide (line 142) ... ignored
test src/lib.rs - guide (line 176) ... ignored
test src/lib.rs - guide (line 214) ... ignored
test src/poll.rs - poll::Poll (line 129) ... ignored
test src/poll.rs - poll::Poll (line 33) ... ignored
test src/poll.rs - poll::Poll::poll (line 263) ... ignored
test src/poll.rs - poll::Registry::deregister (line 573) ... ignored
test src/poll.rs - poll::Registry::register (line 428) ... ignored
test src/poll.rs - poll::Registry::reregister (line 506) ... ignored
test src/token.rs - token::Token (line 20) ... ignored
test src/waker.rs - waker::Waker (line 37) ... ignored

test result: ok. 5 passed; 0 failed; 16 ignored; 0 measured; 0 filtered out; finished in 0.97s

```


soo.. I guess close issue waiting on folks to update :-)

**Comment by Thomasdezeeuw** [comment]
Building Mio without features basically builds everywhere because it's a non-implementation (see https://github.com/tokio-rs/mio/tree/58e589fddeac03453ffe60718c64e2a582c12275/src/sys/shell). But once you enable features we need actual OS support.

So what system calls does Haiku have that we can use? Note that `poll` and `select` are not good enough.

**Comment by kallisti5** [comment]
Weird.. to help me better determine if alternatives exist.. why are poll / select not good enough? select isn't polling and doesn't really waste CPU time.

We have a pretty unique  wait_for_objects call which might be a stand-in for epoll / kqueue:

https://github.com/haiku/haiku/blob/master/headers/os/kernel/OS.h#L634

It's experimental however.. and documentation could be (a lot) better.

**Comment by kallisti5** [comment]
Just found a simple usage example of wait_for_events here:

```
https://git.haiku-os.org/haiku/tree/src/servers/power/power_daemon.cpp#n118
```

**Comment by Thomasdezeeuw** [comment]
The problem with `poll` and `select` is that you need all the file descriptors to make the call. With `epoll`/`kqueue` the OS allows us to register an interest in a file descriptor, which is done in [`mio::Registry::register`](https://docs.rs/mio/0.7.9/mio/struct.Registry.html#method.register). That allows us to keep no user-space state about what file descriptor we're actually polling.

If were to support `poll`/`select` we need a copy of all file descriptors that are registered. Which is possible, but would require a lot of work to support in Mio.

**Comment by Thomasdezeeuw** [comment]
> We have a pretty unique wait_for_objects call which might be a stand-in for epoll / kqueue:
> 
> https://github.com/haiku/haiku/blob/master/headers/os/kernel/OS.h#L634

Am I correct in thinking you still need to pass all the `object_wait_info` objects you're interested in for `wait_for_objects`? Because if so it's not good enough, see my previous comment.

**Comment by kallisti5** [comment]
Hm. Yeah. You build the list of objects and operate on them (aka stateless)
tldr; the design of mio assumes stateful file descriptor monitoring. 

We have a GSoC session coming up.. one topic is adding io_uring support to Haiku.  This seemingly represents the "future" of fd monitoring on Linux and elsewher.  Any thoughts on the potential for mio to move to that model at some point in the future?

I was linked a great FOSDEM video here:
https://fosdem.org/2021/schedule/event/file_descriptor_monitoring/

**Comment by kallisti5** [comment]
Ok.  I went ahead and opened this ticket based on some discussions around this stuff:
  https://dev.haiku-os.org/ticket/16846

We generally only implement POSIX stuff, but it feels like we're lacking here.

as for us targeting a io_uring API in the future, I feel like any discussions around io_uring would duplicate #923 so I won't bring them up :grin:     Part of 16846 above would be offering a epoll/kqueue compatibility API leveraging a native io_uring implementation.

With that said.. feel free to close this one if you deem it impractical for the moment.  Feel free to point any future people complaining about mio not working under Haiku here :laughing:  

A *lot* of crates eventually depend on mio though tokio.. I guess i'm going to have to avoid leveraging anything that uses tokio for a while.   Thanks!

**Comment by Thomasdezeeuw** [comment]
Feel free to reopen or create a new issue once something like https://dev.haiku-os.org/ticket/16846 is possible on Haiku.

**Comment by waddlesplash** [comment]
> https://dev.haiku-os.org/ticket/16846 is possible on Haiku.

This has been done by implementing a (subset of) kqueue, and it's now available on nightly builds, so this issue can be reopened.

**Comment by Thomasdezeeuw** [comment]
> > https://dev.haiku-os.org/ticket/16846 is possible on Haiku.
> 
> This has been done by implementing a (subset of) kqueue, and it's now available on nightly builds, so this issue can be reopened.

Does Mio test suite pass on Haiku with kqueue?

**Comment by waddlesplash** [comment]
I don't know, because I don't have a Rust setup on Haiku at the moment. @nielx or @kallisti5 may be more qualified to investigate.

**Comment by pheki** [comment]
Just in case it's helpful in the future, I had prototyped running mio tests on a haiku vm in github actions on https://github.com/tokio-rs/mio/pull/1602#issuecomment-1605237018

Commits and diff on GitHub: https://github.com/tokio-rs/mio/compare/1387b178629f80de65b112c416ce41f047447e9e...d4eef2f3735e8eb77b7f89586cbb0a5e69d435cf

If the commits eventually disappear (I deleted my fork), here's the full ci.yml and diff:

[ci.yml](https://github.com/tokio-rs/mio/files/12223500/ci-yml.txt)
[ci-diff.txt](https://github.com/tokio-rs/mio/files/12223502/ci-diff.txt)

**Comment by Thomasdezeeuw** [comment]
For anyone still watching this issue. Mio v1 will support Haiku using the `poll(2)` and `pipe(2)` implementations. I don't know the state of `epoll`/`kqueue` of Haiku, I find can't find any manual for these kind of system calls other than a reference to the Posix standard. If an API like `epoll`/`kqueue` is ever support by Haiku we should switch to that as it would be far more efficient.

**Comment by nielx** [comment]
There is a basic kqueue implementation, so it is worth investigating if it has all the features to switch over. I'll add it to my list.

**Comment by Fancy2209** [comment]
Has there been any update on this now that kqueue is a thing? I'd love to get ruffle working on haiku and tokio is what is currently preventing it
